dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
+        {
+            if (dfd.exists(lfn, user))
+                throw MakeStringException(-1, "Super file '%s' already exists!", lfn);
+
+            // It seems XML is fine to create  super file
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(t);
+
+            superfile->validate();
```

Comment:
I think there was an earlier comment , that validate would be better as a common method (virtual of IDistributedFile), for the super to do what you've implemented and for the base File version to do what it needs to do for file verification that you do below, i.e. if (file->existsPhysicalPartFiles(0)) ..